apps/homescreen/js/app.js

```diff
@@ -164,141 +164,44 @@ const SETTINGS_VERSION = 0;
-              () => {},
-              (e) => {
-                console.error('Error removing unknown app metadata entry', e);
-              });
-          }
-        }
-        this.startupMetadata = [];
-        this.storeAppOrder();
-
-        // All asynchronous loading has finished
-        window.performance.mark('fullyLoaded');
-      });
-    };
-
+    // Populate apps and bookmarks asynchronously
     this.startupMetadata = [];
     this.iconsToRetry = [];
+    this.pendingIcons = {};
     this.metadata = new HomeMetadata();
     this.bookmarks = new Datastore('bookmarks_store');
-    this.pages = new Pages();
-
-    // Load metadata, then populate apps. If metadata loading fails,
-    // populate apps anyway - it means they'll be in the default order
-    // and their order won't save, but it's better than showing a blank
-    // screen.
-    // If this is the first run, get the app order from the first-run script
-    // after initialising the metadata database.
-    new Promise((resolve, reject) => {
-      this.metadata.init().then(() => {
-        if (this.firstRun) {
-          resolve();
-          return;
-        }
 
-        this.metadata.getAll().then((results) => {
-          this.startupMetadata = results;
-          resolve();
+    // Make sure icons isn't doing lots of unnecessary work while we're
+    // loading the first screen of apps.
+    this.visualLoadComplete = false;
+    this.icons.freeze();
+    this.icons.classList.add('loading');
+
+    Promise.all([
+      // Load app metadata. If metadata loading fails, continue populating apps
+      // anyway - it means they'll be in the default order and their order
+      // won't save, but it's better than showing a blank screen.
+      // If this is the first run, get the app order from the first-run script
+      // after initialising the metadata database.
+      this.metadata.init().then(this.firstRun ? Promise.resolve() :
+        () => {
+          return this.metadata.getAll(result => {
+            // Process results as they come in
+            this.startupMetadata.push(result);
+            if (this.pendingIcons[result.id]) {
+              this.addAppIcon.apply(this, this.pendingIcons[result.id]);
+              delete this.pendingIcons[result.id];
+            }
+          }).then(Promise.resolve(),
+            e => {
+              console.error('Failed to retrieve metadata entries', e);
+              return Promise.resolve();
+            });
         },
         (e) => {
-          console.error('Failed to retrieve metadata entries', e);
-          resolve();
-        });
-      },
-      (e) => {
-        console.error('Failed to initialise metadata db', e);
-        resolve();
-      });
-    }).then(this.firstRun ?
+          console.error('Failed to initialise metadata db', e);
+          return Promise.resolve();
+        }).then(this.firstRun ?
         LazyLoader.load(['js/firstrun.js'],
           () => {
```

Comment:
Actually, the first-run stuff could go further up instead of the Promise.resolve(), which would cut down on redundant Promise.resolves and this.firstRun comparisons... I'll see about doing that.